Objects/bytearrayobject.c

```diff
@@ -245,26 +247,27 @@ bytearray_resize_lock_held(PyObject *self, Py_ssize_t requested_size)
     }
 
     if (logical_offset > 0) {
-        sval = PyMem_Malloc(alloc);
-        if (sval == NULL) {
-            PyErr_NoMemory();
+        /* Make a new non-offset buffer */
+        PyObject *new = PyBytes_FromStringAndSize(NULL, alloc - 1);
+        if (new == NULL) {
             return -1;
         }
-        memcpy(sval, PyByteArray_AS_STRING(self),
+        memcpy(PyBytes_AS_STRING(new), PyByteArray_AS_STRING(self),
                Py_MIN((size_t)requested_size, (size_t)Py_SIZE(self)));
-        PyMem_Free(obj->ob_bytes);
+        Py_DECREF(obj->ob_bytes_head);
+        bytearray_set_bytes(obj, new, size);
+        obj->ob_bytes[size] = '\0'; /* Trailing null byte */
+
+        return 0;
     }
-    else {
-        sval = PyMem_Realloc(obj->ob_bytes, alloc);
-        if (sval == NULL) {
-            PyErr_NoMemory();
-            return -1;
-        }
+
+    /* Too small, grow allocation */
+    if (_PyBytes_Resize(&obj->ob_bytes_head, alloc - 1) < 0) {
```

Comment:
behavior change: `_Py_Bytes_Resize` behaves differently than `PyMem_Realloc` here. `_PyBytes_Resize` can _change_ the location of the buffer. That means the address returned from `PyByteArray_AS_STRING` may change on any resize. Previously it changed only when went from empty <-> non-empty bytearray (empty points to a constant storage).

https://docs.python.org/3/c-api/memory.html#c.PyMem_Realloc